### PR TITLE
WIP Rework time handling

### LINKED ISSUE #7: Remove clock_gettime()

### PR DESCRIPTION
**Pull request by xrmx**
Switch TSClock and TSDuration to backed by time::Instant and time::Duration.

Some code did get involded in order to keep the u64 interface. In the process clock_is_null got removed.

Fix #7

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -2039,8 +2039,7 @@ unsafe extern "C" fn ts_parser__advance(
         if (*self_0).operation_count == 0 as libc::c_int as libc::c_uint
             && (!(*self_0).cancellation_flag.is_null()
                 && atomic_load((*self_0).cancellation_flag) != 0
-                || !clock_is_null((*self_0).end_clock)
-                    && clock_is_gt(clock_now(), (*self_0).end_clock) as libc::c_int != 0)
+                || clock_is_gt(clock_now(), (*self_0).end_clock) as libc::c_int != 0)
```
Comment (xrmx) [inline src/parser.rs]: i can avoid the cast to c type here i guess?

File: src/parser.rs [hunk 2]
```diff
@@ -2585,7 +2584,7 @@ pub unsafe extern "C" fn ts_parser_set_cancellation_flag(
     (*self_0).cancellation_flag = flag as *const size_t;
 }
 #[no_mangle]
-pub unsafe extern "C" fn ts_parser_timeout_micros(mut self_0: *const TSParser) -> uint64_t {
+pub unsafe extern "C" fn ts_parser_timeout_micros(mut self_0: *const TSParser) -> u64 {
```
Comment (xrmx) [inline src/parser.rs]: not sure i should have changed the type

File: src/api.rs [hunk 3]
```diff
@@ -1228,50 +1223,23 @@ pub unsafe extern "C" fn ts_decode_utf8(
...
-    result.tv_sec = (result.tv_sec as libc::c_ulong)
-        .wrapping_add(duration.wrapping_div(1000000 as libc::c_int as libc::c_ulong))
-        as __time_t as __time_t;
-    result.tv_nsec = (result.tv_nsec as libc::c_ulong).wrapping_add(
-        duration
-            .wrapping_rem(1000000 as libc::c_int as libc::c_ulong)
-            .wrapping_mul(1000 as libc::c_int as libc::c_ulong),
-    ) as __syscall_slong_t as __syscall_slong_t;
-    return result;
+pub extern "C" fn clock_null() -> Option<TSClock> {
```
Comment (xrmx) [inline src/api.rs]: I think i can kill this

**Review by lu-zero — APPROVED** [review]
if this piece of code is used only internally it could be simplified further, but that can happen later.

Looks good to me.

**Review by lu-zero — APPROVED** [review]
if this piece of code is used only internally it could be simplified further, but that can happen later.

Looks good to me.

File: src/api.rs (hunk 3 above)
Comment (lu-zero) [reply src/api.rs]: Yes

File: src/parser.rs (hunk 1 above)
Comment (xrmx) [reply src/parser.rs]: Removed

File: src/parser.rs (hunk 2 above)
Comment (xrmx) [reply src/parser.rs]: reverted to original type

**Comment by xrmx** [comment]
Updated after review and pushed

File: src/api.rs [hunk 4]
```diff
@@ -66,14 +67,8 @@ pub static mut LENGTH_MAX: Length = {
...
 
-pub type TSDuration = uint64_t;
-// POSIX with monotonic clock support (Linux)
-// * Represent a time as a monotonic (seconds, nanoseconds) pair.
-// * Represent a duration as a number of microseconds.
-//
-// On these platforms, parse timeouts will correspond accurately to
-// real time, regardless of what other processes are running.
-pub type TSClock = timespec;
+pub type TSDuration = Duration;
```
Comment (xrmx) [inline src/api.rs:70]: Nope, we can't change the types :broken_heart:

**Pull request closed without merging**